data_processing/eia.py

```diff
@@ -0,0 +1,2999 @@
+        'HK': 'Hydrokinetic Other',
+        'HY': 'Hydraulic Turbine',
+        'IC': 'Internal Combustion Engine',
+        'PS': 'Pumped Storage',
+        'OT': 'Other',
+        'ST': 'Steam Turbine',
+        'PV': 'Photovoltaic',
+        'WT': 'Wind Turbine Onshore',
+        'WT': 'Wind Turbine Offshore'
+    }
+
+    return mover
+
+
+def generation(year: str):
+    '''
+    This function generates a nested dictionary of the Universities
+    present in the EIA spreadsheet and their generation amounts
+    by each fuel/mover type combination.
+
+    Parameters:
+    -----------
+    year: str
+        This is the string for the year being analyzed.
+
+    Returns:
+    --------
+    final: dict
+        Nested dictionary of the Universities present in the EIA
+        spreadsheet and their generation amounts in MWhr by each fuel/mover
+        type combination.
+
+    '''
+
+    if not isinstance(year, str):
+
+        raise TypeError(
+            "Argument 'year' must be of type 'str'."
+        )
+
+    allowed_years = [
+        '2008',
+        '2009',
+        '2010',
+        '2011',
+        '2012',
+        '2013',
+        '2014',
+        '2015',
+        '2016',
+        '2017',
+        '2018'
+    ]
+
+    if year not in allowed_years:
+
+        raise ValueError(
+            f"Not a valid year. Possibilities include: {*allowed_years,}."
+        )
+
+    headers = {
+        '2008': 7,
+        '2009': 7,
+        '2010': 7,
+        '2011': 5,
+        '2012': 5,
+        '2013': 5,
+        '2014': 5,
+        '2015': 5,
+        '2016': 5,
+        '2017': 5,
+        '2018': 5
+    }
+
+    netGen = {
+        '2008': 'NET GENERATION (megawatthours)',
+        '2009': 'NET GENERATION (megawatthours)',
+        '2010': 'NET GENERATION (megawatthours)',
+        '2011': 'Net Generation (Megawatthours)',
+        '2012': 'Net Generation (Megawatthours)',
+        '2013': 'Net Generation (Megawatthours)',
+        '2014': 'Net Generation (Megawatthours)',
+        '2015': 'Net Generation (Megawatthours)',
+        '2016': 'Net Generation (Megawatthours)',
+        '2017': 'Net Generation (Megawatthours)',
+        '2018': 'Net Generation (Megawatthours)'
+    }
+
+    generation = pd.read_csv((path + year + '.csv'),
+                             header=headers[year],
+                             usecols=['Operator Name',
+                                      'NERC Region',
+                                      'NAICS Code',
+                                      'Reported Prime Mover',
+                                      'Reported Fuel Type Code',
+                                      str(netGen[year])])
+
+    generation = generation.loc[generation['NAICS Code'] == 611]
+
+    generation['University'] = generation['Operator Name'].astype(
+        str) + ' ' + generation['NERC Region']
+
+    generation = generation.drop(
+        ['Operator Name', 'NERC Region'],
+        axis=1
+    )
```

Comment:
You're essentially renaming a column by creating a new column and dropping two others. I couldn't find where "NERC Region" came up in your analysis (it wasn't broken up by region anywhere) so we can probably leave it alone.

This is done easily in pandas with the following
```suggestion
    generation = generation.loc[generation['NAICS Code'] == 611]

    generation.rename(columns={'Operator Name':'University'}, inplace=True)

```

Reply:
Nice use of masking to select the "NCAIS code" :smile:

Reply:
So I decided to include the NERC region because there were a few instances I noticed where the Operator is the same University name, but its actually a different institution. For instance "University of Illinois" is listed twice in all of the EIA 293s, but upon further inspection one of them is the University of Illinois Chicago and one is Urbana-Champaign. The only distinguishing column was the NERC region. If I removed the region it combined the numbers and University of Illinois became the largest producer.
Do you have any better solutions to combat this problem besides hard-coding all the duplicates?

Reply:
I made the suggestion because it wasn't obvious to me why the NERC region
was necessary. Based on your description of the problem it is necessary to
distinguish institutions, so I think this solution is fine. I might do:

```python
universities = generation['NAICS Code'] == 611
generation = generation.loc[universities]
generation['University'] = generation['Operator Name'] + ' '  + generation['NERC Region']
generation.drop(['Operator Name', 'NERC Region'], axis=1, inplace=True)
```
Since they're already strings you can just use the ``+`` operator. I added
a line for masking (the first line) because it's clearer
to the reader why you're selecting that code. (At least, I'm assuming
they're universities).